### PR TITLE
Data preprocessing 

### PR DESCRIPTION
**Pull request by AlexLipp**
This PR adds some basic pre-processing functions to funmixer. In particular it adds functions to: 

1. Check the validity of an input D8 raster with respect to boundary conditions and values 
2. Fix the boundary conditions automatically
3. "Snap" sample sites to drainage to allow for sensible building of SampleNetwork objects (raised in Issue #23). 

These functions are now flagged in the README and an example of use is now included in the examples directory.

### CODE REVIEW

File: funmixer/d8processing.py [hunk 1]
```diff
@@ -0,0 +1,377 @@
+"""
+This module contains functions for (pre)processing D8 flow direction grids and snapping sample sites to drainage networks.
+"""
+
+from osgeo import gdal
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+from typing import Tuple, List, Dict
```
Comment (r-barnes) [inline funmixer/d8processing.py:10]: If we commit to Python 3.10+ then we can use `tuple`, `list`, and `dict` without the import.

File: funmixer/d8processing.py [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+    """
+    arr, ds = read_geo_file(flowdirs_filename)
+    arr[0, :] = 0
+    arr[-1, :] = 0
+    arr[:, 0] = 0
+    arr[:, -1] = 0
+
+    # Save the new file with suffix [original_filename]_fix_bounds.tif, overwriting the original file extension which may not be .tif
+    # Split the filename into the base and extension
+    base, _ = flowdirs_filename.rsplit(".", 1)
```
Comment (r-barnes) [inline funmixer/d8processing.py]: If would be better to use, eg, `pathlib.Path(filename).stem` to get this. Prefer `pathlib` for any manipulation of filenames/filepaths.

**Review by r-barnes — APPROVED** [review]

File: funmixer/d8processing.py (hunk 2 above)
Comment (AlexLipp) [reply funmixer/d8processing.py]: I've resolved this now to use pathlib

**Comment by AlexLipp** [comment]
Modified to include pathlib for filename manipulation

**Pull request merged**